File: Hot 100/Arrays/0442_find_all_duplicates_in_an_array.java
// hashset solution
class Solution {
    public List<Integer> findDuplicates(int[] nums) {
        Set<Integer> hashset = new HashSet<>();
        int n = nums.length;
        List<Integer> ans = new ArrayList<>();
        for (int i = 0; i < n; i++) {
            if (!hashset.contains(nums[i])) {
                hashset.add(nums[i]);
            } else {
                ans.add(nums[i]);
            }
        }
        return ans;
    }
}

File: Hot 100/Arrays/0442_find_all_duplicates_in_an_array.py
class Solution:
    def findDuplicates(self, nums: List[int]) -> List[int]:
        n = len(nums)
        for i in range(n):
            while nums[i] != nums[nums[i] - 1]:
                nums[nums[i] - 1], nums[i] = nums[i], nums[nums[i] - 1]
        ans = []
        for i in range(n):
            if nums[i] != i + 1:
                ans.append(nums[i])
        return ans
